velero-csi-plugin/util.go

```diff
@@ -0,0 +1,143 @@
+	}
+	if pvc.Status.Phase != corev1api.ClaimBound {
+		// TODO: confirm if this PVC should be snapshotted if it has no PV bound
+		return nil, errors.Errorf("PVC %s/%s is in phase %v and is not bound to a volume", pvc.Namespace, pvc.Name, pvc.Status.Phase)
+	}
+	pvName := pvc.Spec.VolumeName
+	pv, err := corev1.PersistentVolumes().Get(pvName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get PV %s for PVC %s/%s", pvName, pvc.Namespace, pvc.Name)
+	}
+	return pv, nil
+}
+
+func getPodsUsingPVC(pvcNamespace, pvcName string, corev1 corev1client.PodsGetter) ([]corev1api.Pod, error) {
+	podsUsingPVC := []corev1api.Pod{}
+	for {
+		podList, err := corev1.Pods(pvcNamespace).List(metav1.ListOptions{})
+		if err != nil {
+			return podsUsingPVC, err
+		}
+
+		for _, p := range podList.Items {
+			for _, v := range p.Spec.Volumes {
+				if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvcName {
+					podsUsingPVC = append(podsUsingPVC, p)
+				}
+			}
+		}
+		if podList.Continue == "" {
+			break
+		}
+	}
+
+	return podsUsingPVC, nil
+}
+
+func getPodVolumeNameForPVC(pod corev1api.Pod, pvcName string) (string, error) {
+	for _, v := range pod.Spec.Volumes {
+		if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvcName {
+			return v.Name, nil
+		}
+	}
+	return "", errors.Errorf("Pod %s/%s does not use PVC %s/%s", pod.Namespace, pod.Name, pod.Namespace, pvcName)
+}
+
+func getPodVolumesUsingRestic(pod corev1api.Pod) []string {
+	resticAnnotation := pod.Annotations[resticPodAnnotation]
+	if resticAnnotation == "" {
+		return []string{}
+	}
+	return strings.Split(pod.Annotations[resticPodAnnotation], ",")
+}
+
+func contains(slice []string, key string) bool {
+	for _, i := range slice {
+		if i == key {
+			return true
+		}
+	}
+	return false
+}
+
+func isPVCBackedUpByRestic(pvcNamespace, pvcName string, podClient corev1client.PodsGetter) (bool, error) {
+	pods, err := getPodsUsingPVC(pvcNamespace, pvcName, podClient)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	for _, p := range pods {
+		resticVols := getPodVolumesUsingRestic(p)
+		if len(resticVols) > 0 {
+			volName, err := getPodVolumeNameForPVC(p, pvcName)
+			if err != nil {
+				return false, err
+			}
+			if contains(resticVols, volName) {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func getVolumeSnapshotClassForStorageClass(provisioner string, snapshotGetter snapshotter.VolumeSnapshotClassesGetter) (*snapshotv1beta1api.VolumeSnapshotClass, error) {
+	for {
+		snapshotClasses, err := snapshotGetter.VolumeSnapshotClasses().List(metav1.ListOptions{})
+		if err != nil {
+			return nil, errors.Wrap(err, "error listing volumesnapshot classes")
+		}
+		for _, sc := range snapshotClasses.Items {
+			if sc.Driver == provisioner {
+				return &sc, nil
+			}
+		}
+		if snapshotClasses.Continue == "" {
```

Comment:
this might be a little confusing given that we're not actually using a limit on the List call - let's simplify to just do a single list and iterate through it, rather than having the outer loop.